daemon/algod/api/server/v2/handlers.go

```diff
@@ -1114,6 +1115,42 @@ func (v2 *Handlers) GetApplicationByID(ctx echo.Context, applicationID uint64) e
 	return ctx.JSON(http.StatusOK, response)
 }
 
+// GetApplicationBoxes returns the box names of an application
+// (GET /v2/applications/{application-id}/boxes)
+func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64, params generated.GetApplicationBoxesParams) error {
+	appIdx := basics.AppIndex(applicationID)
+	ledger := v2.Node.LedgerForAPI()
+	lastRound := ledger.Latest()
+	keyPrefix := logic.MakeBoxKey(appIdx, "")
+
+	var maxReturnNum uint64
+	if params.Max == nil || *params.Max == 0 {
```

Comment:
nit: Could this add an or condition such that:
```suggestion
	if params.Max == nil || *params.Max == 0 || *params.Max > v2.Node.Config().MaxAPIBoxPerApplication {
```

 It might be a little less readable but more concise.

Just an observation, but it looks like the `AccountInfo` endpoint returns an [error with some extra data](https://github.com/algorand/go-algorand/blob/b4e7edb7d3d2599cb316671fb85dde324e4f5a3a/daemon/algod/api/server/v2/handlers.go#L310) - maybe something that can be done here too.

Reply:
good point, I adapt to your suggestion.

Reply:
That note about the error seems important. It seems that when we return truncated results in the other endpoint, we say so.

Reply:
I agree, so I added the note about query number exceeds inside v2 log in latest commit: https://github.com/algorand/go-algorand/pull/4122/commits/d841a9a892bd0db8ac07817dac5d35f1da044583

Reply:
After closer inspection, it seems like the `AccountInfo` endpoint doesn't even truncate the results, it just reports an 400 error with the number of resources that the account is associated with:

https://github.com/algorand/go-algorand/blob/b4e7edb7d3d2599cb316671fb85dde324e4f5a3a/daemon/algod/api/server/v2/handlers.go#L319

I don't have a strong opinion on whether we should truncate or return error. But the other API does return all results if the config max setting is set to 0 as well and could be applicable here (maybe someone would want their node to return all the results?):

https://github.com/algorand/go-algorand/blob/b4e7edb7d3d2599cb316671fb85dde324e4f5a3a/config/localTemplate.go#L439

Reply:
Let's follow that pattern to minimize surprises.

You can fail early by checking the number of boxes in the account object.

And yes, let's let algod runners choose to return all results if they desire.

Reply:
> Let's follow that pattern to minimize surprises.

Double check, the pattern we are going to follow is error early on reporting 400 error in `AccountInformation`?

> let algod runners choose to return all results if they desire.

By passing 0/undefined in `ApplicationBoxesParams` to return *up to `maxAPIBoxNameThreshold`*, not really all of them?

Reply:
By an "algod runner" here I'm talking about the people who configure algod, so they are the ones that set `MaxAPIBoxNameThreshold`. If they set it to 0, they are saying they want their algod to return an arbitrary number of box names to REST callers.

This technique (of checking the account.Boxes field first and bail out if there are too many) will only work for attempts to read all the boxes of an app.  We did not intend (yet) to offer a way to ask for all the boxes that start with some prefix, right?

Reply:
ok, I kinda see how this pattern works now. But the technique of bail out early is only workable if we want to read all boxes of an app, and it is not yet offered.

Reply:
Right, I think you would have to read all the boxes first then exit. If we were to mirror the existing API, I think we should remove the `maxKeyNum` as an input to the ledger/lookup APIs, or (maybe less cleaner) have `maxKeyNum=0` to return all the keys.

Reply:
I will keep `maxKeyNum` as an argument, but I definitely think we should let `maxKeyNum = 0` returns everything, which follows an old design.

Reply:
Thanks for handling! Resolving this